src/types/middleware.ts

```diff
@@ -0,0 +1,28 @@
+import { EffectScope } from 'vue'
+import { MaybePromise } from '@/types/utilities'
+
+type Route = {
+  name: string,
+  params: Record<string, unknown>,
+  query: unknown,
+  hash: string,
+}
+
+interface RouteState {
+
+}
+
+interface RouteRejection {
+  NotFound: unknown,
+}
```

Comment:
This is an interface so users can add their own rejections types at implementation. Each rejection type will be able to render a component specified when creating the router.